### ISSUE TITLE
Wanted behavior for `deepcopy(dim_tag)`?

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#860]
Do we want:
```
tag == deepcopy(tag)
```
or not?

In returnn_common (and similarly in PyTorch), it is common do do sth like this:
```
class MultiLayer(Module):
  def __init__(self, out_dim: nn.Dim, num_layers: int, layer: Optional[Module] = None):
    if not layer:
      layer = nn.Linear(out_dim)
    self.layers = nn.Sequential(deepcopy(layer) for _ in range(num_layers)
```

That `deepcopy` on such layer has a reference to the `out_dim`.

The current behavior has `tag != deepcopy(tag)`, although I think this is actually buggy for other reasons.

Related is also how we actually define the equality of dim tags (#634). Although, maybe not so much, as we can keep the `deepcopy` logic as we want it.

I'm somewhat undecided about what behavior we want, or we expect. I tend to prefer `tag == deepcopy(tag)`, also because this simplifies the code in returnn_common.

**Comment by albertz** [comment]
@Zettelkasten opinions?

**Comment by albertz** [comment]
To get `tag == deepcopy(tag)`, I think we can simply add this to `Dim`:
```
def __copy__(self):
  return self

def __deepcopy__(self, memo):
  return self
```

**Comment by Zettelkasten** [comment]
Hm I'm also not sure.
Would it be too bad if we had some `dim.copy()` with a different behavior than `dim.__copy__()`?
I think if I explicitly call `dim.copy()` in a config (e.g. to do an outer product), I'd expect it to be different to the old one.
But, at the same time, it would be super confusing to have two dims with the same name, so I would always rename it when I copy it like this (i.e. have some `dim.copy(dim.description + '-other')` or so).

Otherwise I definitely agree that `deepcopy(module)` is a very valid use case where you do not want the dim tag to be changed.

What do you think of having `__copy__` and `__deepcopy__` such that it does nothing, but to add `Dim.copy(description: str)` which actually returns a copy (and furthermore asserts that you do not create two dim tags with the same description).
Would that be too confusing?

**Comment by albertz** [comment]
> Hm I'm also not sure. Would it be too bad if we had some `dim.copy()` with a different behavior than `dim.__copy__()`? I think if I explicitly call `dim.copy()` in a config (e.g. to do an outer product), I'd expect it to be different to the old one. But, at the same time, it would be super confusing to have two dims with the same name, so I would always rename it when I copy it like this (i.e. have some `dim.copy(dim.description + '-other')` or so).

Yes exactly. Having a separate explicit `copy` function would make it clear that this is really copying a separate instance (otherwise there is anyway no point in copying it). Maybe it would have a mandatory `description` argument because as you say, it would be confusing with the same description. But also, I cannot really think of any use case of such `copy` function right now anyway.

> What do you think of having `__copy__` and `__deepcopy__` such that it does nothing, but to add `Dim.copy(description: str)` which actually returns a copy (and furthermore asserts that you do not create two dim tags with the same description). Would that be too confusing?

I think this is fine. But actually we don't really need the `copy` function for now. I think I would leave that away, unless we really need it at some point. In `__copy__` and `__deepcopy__`, there should be a good comment and explanation on the behavior of course, maybe just with a link to this issue.

**Comment by Zettelkasten** [comment]
Okay, sure, let's add `copy` separately.
I actually could use `dim_other = dim.copy(dim.description + 'other')` in my config though, I have it e.g. here where I want to create a diagonal mask where `i < j`, I do this:
```
other_window_offset_dim = Dim(
  kind=Dim.Types.Spatial, description="other-window-offset", dimension=window_offset_dim.dimension)
```
and then range into both `window_offset_dim` and `other_window_offset_dim` before comparing the two.

**Comment by albertz** [comment]
Actually we already have this.
```
def copy(self, kind=None):
  """
  :param Entity|None kind: if set, overwrites self.kind
  :return: copy, maybe as new kind. setting same_as to self
  :rtype: Dim
  """
  tag = Dim(
    kind=kind or self.kind, description=self.description,
    dimension=self.dimension, dyn_size_ext=self.dyn_size_ext,
    batch=self.batch,
    src_data=self.src_data, src_axis=self.src_axis)
  tag.same_as = self  # not declare_same_as, none of the extra checks needed
  tag._same_as_tb = traceback.extract_stack()
  return tag
```
I.e. not as you suggested, it sets `same_as = self`.

It is only used in one single place though, in `Data.copy_add_dim_by_tag`:
```
# Note: if dim_tag is feature, but we are sparse, we just make it spatial
if self.sparse and dim_tag.is_feature_dim():
  dim_tag = dim_tag.copy(kind=Dim.Types.Spatial)
```
(Not sure if this is really needed anyway...)

Maybe it should have a new mandatory option `same_as_self: bool` to avoid any confusion.
